File: Xliffie/Glossary/CustomGlossaryDatabase.m
//
//  CustomGlossaryDatabase.m
//  Xliffie
//
//  Created by b123400 on 2024/02/08.
//  Copyright © 2024 b123400. All rights reserved.
//

#import "CustomGlossaryDatabase.h"
#import <sqlite3.h>
#import "Utilities.h"
#import "CHCSVParser.h"

@implementation CustomGlossaryDatabase {
    sqlite3 *_sqlite;
}

+ (instancetype)shared {
    static CustomGlossaryDatabase *shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shared = [[CustomGlossaryDatabase alloc] init];
    });
    return shared;
}

- (instancetype)init {
    if (self = [super init]) {
        [self open];
        self.notificationEnabled = YES;
    }
    return self;
}

- (NSString *)databasePath {
    NSArray<NSURL *> *documentPaths = [[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask];
    NSString *docPath = [[documentPaths lastObject] path];
    NSString *dbPath = [docPath stringByAppendingPathComponent:@"custom_glossary.db"];
    NSLog(@"custom db path %@", dbPath);
    return dbPath;
}

- (BOOL)open {
    if (_sqlite) return YES;

    sqlite3 *dbConnection = nil;
    int rc = sqlite3_open_v2([self.databasePath UTF8String], &dbConnection, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, NULL);
    if (rc != SQLITE_OK) {
        return NO;
    }
    _sqlite = dbConnection;
    [self query:@"CREATE TABLE IF NOT EXISTS glossary (id INTEGER PRIMARY KEY AUTOINCREMENT, source_locale varchar(255) NULL, target_locale varchar(255) NULL, source text NOT NULL, target text NOT NULL);" withParams:@[]];
    return YES;
}

- (CustomGlossaryRow *)insertWithSourceLocale:(NSString  * _Nullable)sourceLocale
                                 targetLocale:(NSString * _Nullable)targetLocale
                                       source:(NSString *)source
                                       target:(NSString *)target {
    NSArray *insertResult = [self query:@"INSERT INTO glossary (source_locale, target_locale, source, target) VALUES (?, ?, ?, ?) RETURNING id, source_locale, target_locale, source, target" withParams:@[
        sourceLocale ?: [NSNull null],
        targetLocale ?: [NSNull null],
        source ?: @"",
        target ?: @"",
    ]];
    if (self.notificationEnabled) {
        [[NSNotificationCenter defaultCenter] postNotificationName:CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION object:self];
    }
    return [[self rowsToObjects:insertResult] firstObject];
}

- (void)deleteRow:(CustomGlossaryRow *)row {
    [self query:@"DELETE FROM glossary WHERE id = ?" withParams:@[row.id]];
    if (self.notificationEnabled) {
        [[NSNotificationCenter defaultCenter] postNotificationName:CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION object:self];
    }
}

- (void)updateRow:(CustomGlossaryRow *)row {
    [self query:@"UPDATE glossary SET source_locale = ?, target_locale = ?, source = ?, target = ? WHERE id = ?" withParams:@[
        row.sourceLocale ?: [NSNull null],
        row.targetLocale ?: [NSNull null],
        row.source,
        row.target,
        row.id,
    ]];
    if (self.notificationEnabled) {
        [[NSNotificationCenter defaultCenter] postNotificationName:CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION object:self];
    }
}

- (NSArray<CustomGlossaryRow *> *)rowsWithSourceLocale:(NSString * _Nullable)sourceLocale
                                          targetLocale:(NSString * _Nullable)targetLocale
                                                source:(NSString *)source {
    return [self rowsWithSourceLocales:[Utilities fallbacksWithLocale:sourceLocale]
                         targetLocales:[Utilities fallbacksWithLocale:targetLocale]
                                source:source];
    
}

- (NSArray<CustomGlossaryRow *> *)rowsWithSourceLocales:(NSArray<NSString *> *)sourceLocales
                                          targetLocales:(NSArray<NSString *> *)targetLocales
                                                 source:(NSString *)source {
    NSMutableString *sourcePlaceholders = @"".mutableCopy;
    NSMutableString *targetPlaceholders = @"".mutableCopy;
    NSMutableArray *params = [NSMutableArray array];
    for (NSString *l in sourceLocales) {
        if (sourcePlaceholders.length > 0) {
            [sourcePlaceholders appendFormat:@","];
        }
        [sourcePlaceholders appendFormat:@"?"];
        [params addObject:[l.lowercaseString stringByReplacingOccurrencesOfString:@"_" withString:@"-"]];
    }
    for (NSString *l in targetLocales) {
        if (targetPlaceholders.length > 0) {
            [targetPlaceholders appendFormat:@","];
        }
        [targetPlaceholders appendFormat:@"?"];
        [params addObject:[l.lowercaseString stringByReplacingOccurrencesOfString:@"_" withString:@"-"]];
    }
    NSString *sourceCondition = sourceLocales.count == 0 ? @"" : [NSString stringWithFormat:@"REPLACE(LOWER(source_locale), '_', '-') IN (%@) OR", sourcePlaceholders];
    NSString *targetCondition = targetLocales.count == 0 ? @"" : [NSString stringWithFormat:@"REPLACE(LOWER(target_locale), '_', '-') IN (%@) OR", targetPlaceholders];
    NSString *query  = [NSString stringWithFormat:@"SELECT id, source_locale, target_locale, source, target FROM glossary WHERE (%@ source_locale IS NULL) AND (%@ target_locale IS NULL) AND source = ? COLLATE NOCASE", sourceCondition, targetCondition];
    [params addObject:source];
    NSArray *rows = [self query:query withParams:params];
    return [self rowsToObjects:rows];
}

- (NSArray<CustomGlossaryRow *> *)allRows {
    NSArray *rows = [self query:@"SELECT id, source_locale, target_locale, source, target FROM glossary" withParams:@[]];
    return [self rowsToObjects:rows];
}

- (BOOL)doesRowExistWithSourceLocale:(NSString * _Nullable)sourceLocale
                        targetLocale:(NSString * _Nullable)targetLocale
                              source:(NSString *)source
                              target:(NSString *)target {
    NSArray *sourceLocales = [Utilities fallbacksWithLocale:sourceLocale];
    NSArray *targetLocales = [Utilities fallbacksWithLocale:targetLocale];
    NSMutableString *sourcePlaceholders = @"".mutableCopy;
    NSMutableString *targetPlaceholders = @"".mutableCopy;
    NSMutableArray *params = [NSMutableArray array];
    for (NSString *l in sourceLocales) {
        if (sourcePlaceholders.length > 0) {
            [sourcePlaceholders appendFormat:@","];
        }
        [sourcePlaceholders appendFormat:@"?"];
        [params addObject:[l.lowercaseString stringByReplacingOccurrencesOfString:@"_" withString:@"-"]];
    }
    for (NSString *l in targetLocales) {
        if (targetPlaceholders.length > 0) {
            [targetPlaceholders appendFormat:@","];
        }
        [targetPlaceholders appendFormat:@"?"];
        [params addObject:[l.lowercaseString stringByReplacingOccurrencesOfString:@"_" withString:@"-"]];
    }
    NSString *sourceCondition = sourceLocales.count == 0 ? @"" : [NSString stringWithFormat:@"REPLACE(LOWER(source_locale), '_', '-') IN (%@) OR", sourcePlaceholders];
    NSString *targetCondition = targetLocales.count == 0 ? @"" : [NSString stringWithFormat:@"REPLACE(LOWER(target_locale), '_', '-') IN (%@) OR", targetPlaceholders];
    NSString *query  = [NSString stringWithFormat:@"SELECT COUNT(id) FROM glossary WHERE (%@ source_locale IS NULL) AND (%@ target_locale IS NULL) AND source = ? AND target = ? COLLATE NOCASE", sourceCondition, targetCondition];
    [params addObject:source];
    [params addObject:target];
    NSArray *countId = [self query:query withParams:params];
    NSNumber *count = [[countId firstObject] firstObject];
    if ([count isKindOfClass:[NSNumber class]] && [count integerValue] >= 1) {
        return YES;
    }
    return NO;
}

- (NSArray<CustomGlossaryRow *> *)rowsToObjects:(NSArray *)rows {
    NSMutableArray<CustomGlossaryRow*> *results = [NSMutableArray array];
    for (NSArray *columns in rows) {
        CustomGlossaryRow *r = [[CustomGlossaryRow alloc] init];
        r.id = columns[0];
        r.sourceLocale = [columns[1] isKindOfClass:[NSNull class]] ? nil : columns[1];
        r.targetLocale = [columns[2] isKindOfClass:[NSNull class]] ? nil : columns[2];
        r.source = columns[3];
        r.target = columns[4];
        [results addObject:r];
    }
    return results;
}

- (NSArray<NSArray*> *)query:(NSString *)sql withParams:(NSArray * _Nullable)params {
    if (![self open]) return nil;
    sqlite3_stmt *compiledStatement = nil;
    int rc = 0;
    if ((rc = sqlite3_prepare_v2(_sqlite, [sql UTF8String], -1, &compiledStatement, nil)) != SQLITE_OK) {
        NSLog(@"Cannot prepare sql (%d) : %@", rc, sql);
        return nil;
    }
    if (params) {
        for (int i = 0; i < params.count; i++) {
            id param = params[i];
            if ([param isKindOfClass:[NSString class]]) {
                sqlite3_bind_text(compiledStatement, i + 1, [(NSString*)param UTF8String], -1, SQLITE_TRANSIENT);
            } else if ([param isKindOfClass:[NSNumber class]]) {
                NSNumber *num = param;
                sqlite3_bind_int64(compiledStatement, i + 1, [num longLongValue]);
            } else if ([param isKindOfClass:[NSNull class]]) {
                sqlite3_bind_null(compiledStatement, i + 1);
            }
        }
    }
    NSMutableArray *result = [NSMutableArray array];
    while (sqlite3_step(compiledStatement) == SQLITE_ROW) {
        NSMutableArray *row = [NSMutableArray array];
        for (int i = 0; i < sqlite3_column_count(compiledStatement); i++) {
            int colType = sqlite3_column_type(compiledStatement, i);
            id value;
            if (colType == SQLITE_TEXT) {
                const char *col = (const char *)sqlite3_column_text(compiledStatement, i);
                value = [[NSString alloc] initWithUTF8String:col];
            } else if (colType == SQLITE_INTEGER) {
                int col = sqlite3_column_int(compiledStatement, i);
                value = [NSNumber numberWithInt:col];
            } else if (colType == SQLITE_FLOAT) {
                double col = sqlite3_column_double(compiledStatement, i);
                value = [NSNumber numberWithDouble:col];
            } else if (colType == SQLITE_NULL) {
                value = [NSNull null];
            } else {
                NSLog(@"%s Unknown data type.", __FUNCTION__);
            }
            // Add value to row
            [row addObject:value];
            value = nil;
        }
        // Add row to array
        [result addObject:row];
    }
    sqlite3_finalize(compiledStatement);
    return result;
}

#pragma mark - CSV

- (NSProgress *)exportToFile:(NSString *)path withTotalCount:(int64_t)total callback:(void (^)(NSError *error))callback {
    if (![self open]) return nil;
    
    NSProgress *progress = [NSProgress progressWithTotalUnitCount:total];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        CHCSVWriter *writer = [[CHCSVWriter alloc] initForWritingToCSVFile:path];
        [writer writeLineOfFields:@[@"source_locale", @"target_locale", @"source", @"target"]];
        
        NSString *sql = @"SELECT source_locale, target_locale, source, target FROM glossary";
        
        sqlite3_stmt *compiledStatement = nil;
        int rc = 0;
        if ((rc = sqlite3_prepare_v2(_sqlite, [sql UTF8String], -1, &compiledStatement, nil)) != SQLITE_OK) {
            NSLog(@"Cannot prepare sql (%d) : %@", rc, sql);
            callback([NSError errorWithDomain:@"net.b123400.xliffie.error" code:0 userInfo:@{
                NSLocalizedDescriptionKey: @"Cannot prepare export SQL"
            }]);
            return;
        }
        int completeCount = 0;
        while (sqlite3_step(compiledStatement) == SQLITE_ROW) {
            for (int i = 0; i < sqlite3_column_count(compiledStatement); i++) {
                int colType = sqlite3_column_type(compiledStatement, i);
                if (colType == SQLITE_TEXT) {
                    const char *col = (const char *)sqlite3_column_text(compiledStatement, i);
                    id value = [[NSString alloc] initWithUTF8String:col];
                    [writer writeField:value];
                } else if (colType == SQLITE_NULL) {
                    [writer writeField:@""];
                } else {
                    NSLog(@"%s Unknown data type.", __FUNCTION__);
                }
            }
            [writer finishLine];
            completeCount++;
            [progress setCompletedUnitCount:completeCount];
        }
        sqlite3_finalize(compiledStatement);
        callback(nil);
    });
    return progress;
}

- (NSProgress *)importWithFile:(NSURL *)url callback:(void (^)(NSError *error))callback {
    self.notificationEnabled = NO;
    CustomGlossaryImporter *importer = [[CustomGlossaryImporter alloc] init];
    importer.delegate = self;
    return [importer importFromFile:url withCallback:^(NSError * _Nonnull error) {
        self.notificationEnabled = YES;
        callback(error);
        [[NSNotificationCenter defaultCenter] postNotificationName:CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION object:self];
    }];
}

- (void)didReadRow:(CustomGlossaryRow *)row fromImporter:(id)importer {
    [self insertWithSourceLocale:row.sourceLocale
                    targetLocale:row.targetLocale
                          source:row.source
                          target:row.target];
}

@end

File: Xliffie/Glossary/CustomGlossaryDatabase.swift
//
//  CustomGlossaryDatabase.swift
//  Xliffie
//
//  Created by b123400 on 2024/02/08.
//  Copyright © 2024 b123400. All rights reserved.
//

import Foundation
import SQLite3

let CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION = "CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION"

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

extension Notification.Name {
    static let customGlossaryDatabaseUpdated = Notification.Name(CUSTOM_GLOSSARY_DATABASE_UPDATED_NOTIFICATION)
}

@objc class CustomGlossaryDatabase: NSObject, CustomGlossaryImporterDelegate {

    @objc static let shared = CustomGlossaryDatabase()

    @objc var notificationEnabled: Bool = true

    private var sqlite: OpaquePointer?

    override init() {
        super.init()
        open()
        notificationEnabled = true
    }

    deinit {
        if let sqlite = sqlite {
            sqlite3_close(sqlite)
        }
    }

    private var databasePath: String {
        let documentPaths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let docPath = documentPaths.last!.path
        let dbPath = (docPath as NSString).appendingPathComponent("custom_glossary.db")
        NSLog("custom db path %@", dbPath)
        return dbPath
    }

    @discardableResult
    private func open() -> Bool {
        if sqlite != nil { return true }

        var dbConnection: OpaquePointer?
        let rc = sqlite3_open_v2(databasePath, &dbConnection, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, nil)
        if rc != SQLITE_OK {
            return false
        }
        sqlite = dbConnection
        query("CREATE TABLE IF NOT EXISTS glossary (id INTEGER PRIMARY KEY AUTOINCREMENT, source_locale varchar(255) NULL, target_locale varchar(255) NULL, source text NOT NULL, target text NOT NULL);", params: [])
        return true
    }

    @objc @discardableResult
    func insert(withSourceLocale sourceLocale: String?, targetLocale: String?, source: String, target: String) -> CustomGlossaryRow {
        let insertResult = query(
            "INSERT INTO glossary (source_locale, target_locale, source, target) VALUES (?, ?, ?, ?) RETURNING id, source_locale, target_locale, source, target",
            params: [
                sourceLocale as AnyObject? ?? NSNull(),
                targetLocale as AnyObject? ?? NSNull(),
                source as AnyObject,
                target as AnyObject,
            ]
        )
        if notificationEnabled {
            NotificationCenter.default.post(name: .customGlossaryDatabaseUpdated, object: self)
        }
        return rowsToObjects(insertResult).first!
    }

    @objc func deleteRow(_ row: CustomGlossaryRow) {
        query("DELETE FROM glossary WHERE id = ?", params: [row.id ?? NSNull()])
        if notificationEnabled {
            NotificationCenter.default.post(name: .customGlossaryDatabaseUpdated, object: self)
        }
    }

    @objc func updateRow(_ row: CustomGlossaryRow) {
        query(
            "UPDATE glossary SET source_locale = ?, target_locale = ?, source = ?, target = ? WHERE id = ?",
            params: [
                row.sourceLocale as AnyObject? ?? NSNull(),
                row.targetLocale as AnyObject? ?? NSNull(),
                row.source as AnyObject,
                row.target as AnyObject,
                row.id ?? NSNull(),
            ]
        )
        if notificationEnabled {
            NotificationCenter.default.post(name: .customGlossaryDatabaseUpdated, object: self)
        }
    }

    @objc func rows(withSourceLocale sourceLocale: String?, targetLocale: String?, source: String) -> [CustomGlossaryRow] {
        return rows(
            withSourceLocales: Utilities.fallbacks(withLocale: sourceLocale),
            targetLocales: Utilities.fallbacks(withLocale: targetLocale),
            source: source
        )
    }

    private func rows(withSourceLocales sourceLocales: [String], targetLocales: [String], source: String) -> [CustomGlossaryRow] {
        var sourcePlaceholders = ""
        var targetPlaceholders = ""
        var params: [AnyObject] = []

        for l in sourceLocales {
            if !sourcePlaceholders.isEmpty { sourcePlaceholders += "," }
            sourcePlaceholders += "?"
            params.append(l.lowercased().replacingOccurrences(of: "_", with: "-") as AnyObject)
        }
        for l in targetLocales {
            if !targetPlaceholders.isEmpty { targetPlaceholders += "," }
            targetPlaceholders += "?"
            params.append(l.lowercased().replacingOccurrences(of: "_", with: "-") as AnyObject)
        }

        let sourceCondition = sourceLocales.isEmpty ? "" : "REPLACE(LOWER(source_locale), '_', '-') IN (\(sourcePlaceholders)) OR"
        let targetCondition = targetLocales.isEmpty ? "" : "REPLACE(LOWER(target_locale), '_', '-') IN (\(targetPlaceholders)) OR"
        let sql = "SELECT id, source_locale, target_locale, source, target FROM glossary WHERE (\(sourceCondition) source_locale IS NULL) AND (\(targetCondition) target_locale IS NULL) AND source = ? COLLATE NOCASE"
        params.append(source as AnyObject)

        return rowsToObjects(query(sql, params: params))
    }

    @objc func allRows() -> [CustomGlossaryRow] {
        return rowsToObjects(query("SELECT id, source_locale, target_locale, source, target FROM glossary", params: []))
    }

    @objc func doesRowExist(withSourceLocale sourceLocale: String?, targetLocale: String?, source: String, target: String) -> Bool {
        let sourceLocales = Utilities.fallbacks(withLocale: sourceLocale)
        let targetLocales = Utilities.fallbacks(withLocale: targetLocale)
        var sourcePlaceholders = ""
        var targetPlaceholders = ""
        var params: [AnyObject] = []

        for l in sourceLocales {
            if !sourcePlaceholders.isEmpty { sourcePlaceholders += "," }
            sourcePlaceholders += "?"
            params.append(l.lowercased().replacingOccurrences(of: "_", with: "-") as AnyObject)
        }
        for l in targetLocales {
            if !targetPlaceholders.isEmpty { targetPlaceholders += "," }
            targetPlaceholders += "?"
            params.append(l.lowercased().replacingOccurrences(of: "_", with: "-") as AnyObject)
        }

        let sourceCondition = sourceLocales.isEmpty ? "" : "REPLACE(LOWER(source_locale), '_', '-') IN (\(sourcePlaceholders)) OR"
        let targetCondition = targetLocales.isEmpty ? "" : "REPLACE(LOWER(target_locale), '_', '-') IN (\(targetPlaceholders)) OR"
        let sql = "SELECT COUNT(id) FROM glossary WHERE (\(sourceCondition) source_locale IS NULL) AND (\(targetCondition) target_locale IS NULL) AND source = ? AND target = ? COLLATE NOCASE"
        params.append(source as AnyObject)
        params.append(target as AnyObject)

        let countRows = query(sql, params: params)
        if let count = countRows.first?.first as? NSNumber, count.intValue >= 1 {
            return true
        }
        return false
    }

    private func rowsToObjects(_ rows: [[Any]]) -> [CustomGlossaryRow] {
        return rows.map { columns in
            let r = CustomGlossaryRow()
            r.id = columns[0] as? NSNumber
            r.sourceLocale = (columns[1] is NSNull) ? nil : columns[1] as? String
            r.targetLocale = (columns[2] is NSNull) ? nil : columns[2] as? String
            r.source = columns[3] as? String ?? ""
            r.target = columns[4] as? String ?? ""
            return r
        }
    }

    @discardableResult
    private func query(_ sql: String, params: [AnyObject]) -> [[Any]] {
        if !open() { return [] }

        var compiledStatement: OpaquePointer?
        let rc = sqlite3_prepare_v2(sqlite, sql, -1, &compiledStatement, nil)
        if rc != SQLITE_OK {
            NSLog("Cannot prepare sql (%d) : %@", rc, sql)
            return []
        }

        for (i, param) in params.enumerated() {
            if let str = param as? String {
                sqlite3_bind_text(compiledStatement, Int32(i + 1), str, -1, SQLITE_TRANSIENT)
            } else if let num = param as? NSNumber {
                sqlite3_bind_int64(compiledStatement, Int32(i + 1), num.int64Value)
            } else if param is NSNull {
                sqlite3_bind_null(compiledStatement, Int32(i + 1))
            }
        }

        var result: [[Any]] = []
        while sqlite3_step(compiledStatement) == SQLITE_ROW {
            var row: [Any] = []
            for i in 0..<sqlite3_column_count(compiledStatement) {
                let colType = sqlite3_column_type(compiledStatement, i)
                let value: Any
                switch colType {
                case SQLITE_TEXT:
                    let col = sqlite3_column_text(compiledStatement, i)!
                    value = String(cString: col)
                case SQLITE_INTEGER:
                    value = NSNumber(value: sqlite3_column_int(compiledStatement, i))
                case SQLITE_FLOAT:
                    value = NSNumber(value: sqlite3_column_double(compiledStatement, i))
                case SQLITE_NULL:
                    value = NSNull()
                default:
                    NSLog("%s Unknown data type.", #function)
                    value = NSNull()
                }
                row.append(value)
            }
            result.append(row)
        }
        sqlite3_finalize(compiledStatement)
        return result
    }

    // MARK: - CSV Export

    @objc func exportToFile(_ path: String, withTotalCount total: Int64, callback: @escaping (Error?) -> Void) -> Progress {
        let progress = Progress(totalUnitCount: total)

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self, self.open() else {
                callback(NSError(domain: "net.b123400.xliffie.error", code: 0, userInfo: [NSLocalizedDescriptionKey: "Cannot open database"]))
                return
            }

            let writer = CHCSVWriter(forWritingToCSVFile: path)
            writer?.writeLine(ofFields: NSArray(array: ["source_locale", "target_locale", "source", "target"]))

            let sql = "SELECT source_locale, target_locale, source, target FROM glossary"
            var compiledStatement: OpaquePointer?
            let rc = sqlite3_prepare_v2(self.sqlite, sql, -1, &compiledStatement, nil)
            if rc != SQLITE_OK {
                NSLog("Cannot prepare sql (%d) : %@", rc, sql)
                callback(NSError(domain: "net.b123400.xliffie.error", code: 0, userInfo: [NSLocalizedDescriptionKey: "Cannot prepare export SQL"]))
                return
            }

            var completeCount: Int64 = 0
            while sqlite3_step(compiledStatement) == SQLITE_ROW {
                for i in 0..<sqlite3_column_count(compiledStatement) {
                    let colType = sqlite3_column_type(compiledStatement, i)
                    if colType == SQLITE_TEXT {
                        let col = sqlite3_column_text(compiledStatement, i)!
                        writer?.writeField(String(cString: col))
                    } else if colType == SQLITE_NULL {
                        writer?.writeField("")
                    } else {
                        NSLog("%s Unknown data type.", #function)
                    }
                }
                writer?.finishLine()
                completeCount += 1
                progress.completedUnitCount = completeCount
            }
            sqlite3_finalize(compiledStatement)
            callback(nil)
        }

        return progress
    }

    @objc func importWithFile(_ url: URL, callback: @escaping (Error?) -> Void) -> Progress {
        notificationEnabled = false
        let importer = CustomGlossaryImporter()
        importer.delegate = self
        return importer.importFromFile(url, withCallback: { [weak self] error in
            guard let self = self else { return }
            self.notificationEnabled = true
            callback(error)
            NotificationCenter.default.post(name: .customGlossaryDatabaseUpdated, object: self)
        })
    }

    // MARK: - CustomGlossaryImporterDelegate

    func didReadRow(_ row: CustomGlossaryRow, fromImporter importer: CustomGlossaryImporter) {
        insert(withSourceLocale: row.sourceLocale, targetLocale: row.targetLocale, source: row.source, target: row.target)
    }
}
